scripts/verify-knife

```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
+# Verify mmfile $1 (default "set.mm") and discouraged file $2 (default
+# "discouraged") using metamath-knife.
+# Usage:
+# verify-knife [source=set.mm] [discouraged_file=discouraged]
+
+source="${1:-set.mm}"
```

Comment:
Is it possible to start this script outside the metamath folder?  What if one of the submitted files does not exist, say because of a spelling mistake?

Reply:
Yes, many (most?) scripts assume by default that they are being run from the metamath folder, but that's not a requirement.

If the source file does not exist:
```
error: I/O error: The system cannot find the file specified. (os error 2)
--> missing-source:1:1
 |
 |
1 diagnostics issued.
```

If the discouraged file does not exist:
```
0 diagnostics issued.
diff: missing-discouraged: No such file or directory
The file 'missing-discouraged' needs to be regenerated.
```

Both cases exit with status 1.

Reply:
The reason I mentioned this, is that this handling is not obvious.  So it belongs to the introducing description, like
# If one of the files does not exist, the script terminates with exit code 1.

The assumptions the script makes on start up are called pre-conditions,  and guarantees about the state on termination are called post-conditions.  The post-condition here is that the script checks the integrity of the .mm file in some (which?)  respects, and if that test succeeds and the discouraged file matches (in what way? Is its format fixed?), then, and only then, the script terminates with exit code 0.

Both conditions are written down in the introducing comment.

The following is a bit nit-picking, but maybe you are still on the learning side.  The clean-up code rm discouraged.new is duplicated, and that is not necessary:

metamath-knife && diff
retval="$?"
rm discouraged.new
exit $retval

Reply:
I added a description to the script file. I didn't change the cleanup logic because I want the script to match metamath-knife's behavior when a file is missing (exit with status 1) rather than diff's behavior (exit with status 2).

Reply:
This idea is also worth being added as a comment.  It is hard to guess.

You can still use the tactics I depicted to have the clean-up in just one place.  The early out on error strategy of your code is less maintainable in more complex contexts, without something like an atexit at hand.  See  this post https://stackoverflow.com/questions/2129923/how-to-run-a-command-before-a-bash-script-exits for a more general approach.

All my remarks given here (and the days before) are neither approvals nor disapprovals.  So you can (could have) follow or ignore them.  I just tried to be convincing.  The code is now in a good enough shape that it can be committed from my point of view.  Adjustments to follow some advanced code style can be made afterwards.  Folks should now decide which of the two pull requests is preferred.

Reply:
AFAIK, it's up to you @BTernaryTau: can we merge, or do you plan to make other changes to this MR, following @wlammen's proposals ?

(As for me, this version is better than #4209)

Reply:
I'm good to merge it as is at this point.

Reply:
I'll take that as my cue. Done.